landlab/components/species_evolution/base_taxon.py

```diff
@@ -0,0 +1,129 @@
+    species, population). More generally, Taxon subclasses act as approaches to
+    simulate evolution of biologic groups. Taxon can be made of individual
+    organisms, although SpeciesEvolver currently has no built in functionality
+    for individuals.
+
+    This class is intended to be subclassed. Subclasses must implement the
+    properties and methods of this base class that are designated as abstract.
+    The methods must take the same parameters and both the properties and
+    methods must return the values described in their docstrings.
+    """
+    def __init__(self):
+        self._extant = True
+        self._uid = None
+        self._parent = None
+        self._children = []
+
+    def __repr__(self):
+        return '<{}, uid={}>'.format(self.__class__.__name__, self.uid)
+
+    @property
+    def uid(self):
+        """The unique identifier of the taxon.
+
+        The identifier is a unique integer automatically assigned by
+        SpeciesEvolver once the component begins tracking the taxon. It is
+        read-only as it should not be changed once it is assigned.
+        """
+        return self._uid
+
+    @property
+    def extant(self):
+        """The living state of the taxon.
+
+        The taxon lives at the current model time if ``True``. The taxon is
+        extinct as of the current model time if ``False``.
+        """
+        return self._extant
+
+    @extant.setter
+    def extant(self, value):
+        """Set the living state of the taxon."""
+        self._extant = value
+
+    @property
+    def parent(self):
+        """The parent taxon.
+
+        The parent is the taxon object that produced this object. A value of
+        ``None`` indicates no parent taxon.
+        """
+        return self._parent
+
+    @parent.setter
+    def parent(self, value):
+        """Set the parent taxon.
+
+        This method also appends this taxon to the list of children belonging
+        to the parent taxon.
+        """
+        self._parent = value
+
+        if value is not None:
+            value.children.append(self)
+
+    @property
+    def children(self):
+        """The immediate descendents of the taxon.
+
+        The children are the objects produced by this taxon. An empty list
+        indicates no child taxon.
+        """
+        return self._children
+
+    @property
+    @abstractmethod
+    def range_mask(self):
+        """A mask of the taxon geographic extent.
+
+        The range mask is a boolean numpy array where True values indicate
+        where the taxon is located in the model grid associated with a
+        SpeciesEvolver instance.
+
+        This property must be implemented in a subclass.
+        """
+        # pragma: no cover
+
+    @abstractmethod
+    def _evolve(self, dt, stage, record, id):
+        """Run the evolutionary processes of the taxon.
+
+        SpeciesEvolver loops through the evolution processes of extant taxa in
```

Comment:
This implies to me that there can be more than one 'stage' in a 'dt' but if this is the case, is a 'total_number_of_stages' required?

Reply:
Yes, the `_evolve` Taxon method can be called multiple times (each a "stage") in one time step (when run_one_step of SpeciesEvolver is called). Although `total_number_of_stages` parameter is not required.

SpeciesEvolver `run_one_step` and Taxon `_evolve` method are designed in such a way that taxon types can have different numbers of stages. ZoneTaxon requires 2 stages. Future taxon types may have a different number of stages. Its up to the Taxon creator to consider what's going on in the stages of other taxon.

Reply:
So basically `_evolve` is called once per stage, so if there were 7 stages, it would be called 7 times? That was not clear to me. I think the taxon type has to know what stage or stages it evolves at, right? So if a taxon does nothing at stage 1, and stage got a value of 1, then nothing happens. And that would all be controlled by the creator of the subclass, right?

Possibly this would be more clear to me if it stated explicitly that this might not do anything at some stages, depending on what stages the taxon actually evolves at.

Reply:
What you describe is stated in the abstract _evolve method of the base class, exemplified in ZoneTaxon, and will be demoed in a future tutorial.